torchao/quantization/quant_api.py

```diff
@@ -199,3 +201,382 @@ def replace_conv2d_1x1(conv):
+        scales_precision: torch.dtype = torch.float32,
+    ) -> None:
+        super().__init__()
+        # always pad if needed since it becomes a noop at runtime if not needed
+        # self.origin_in_features = in_features
+        assert (
+            in_features % group_size == 0
+        ), f"require in_features:{in_features} % group_size:{group_size} == 0"
+        # in_features = _calc_padded_size_linear_int4(
+        #    in_features, group_size
+        # )
+        self.in_features = in_features
+        self.out_features = out_features
+        assert not bias, "require bias=False"
+        self.group_size = group_size
+        # Precision of the activation which also indicates
+        # output precision of the dynamically quantized linear layer
+        # that his module represents.
+        self.precision = precision
+
+        # currently storing unpacked int8 weights
+        self.register_buffer(
+            "weight",
+            torch.empty((out_features, in_features), dtype=torch.int8),
+        )
+        self.register_buffer(
+            "scales",
+            torch.empty(
+                (out_features, in_features // group_size),
+                dtype=scales_precision,
+            ),
+        )
+        self.register_buffer(
+            "zeros",
+            torch.empty(
+                (out_features, in_features // group_size),
+                dtype=scales_precision,
+            ),
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        input = input.to(self.precision)
+        # padding is removed for perf
+        # input = F.pad(input, pad=(0, self.in_features - self.origin_in_features))
+        return linear_forward_8da4w(
+            input,
+            self.weight,
+            self.scales,
+            self.zeros,
+            self.out_features,
+            self.groupsize,
+            self.precision,
+        )
+
+
+def replace_linear_8da4w(
+    module,
+    group_size,
+    padding_allowed,
+    precision,
+    scales_precision,
+):
+    for name, child in module.named_children():
+        if isinstance(child, nn.Linear):
+            if _check_linear_int4_k(child.in_features, group_size) or padding_allowed:
+                setattr(
+                    module,
+                    name,
+                    Int8DynActInt4WeightLinear(
+                        child.in_features,
+                        child.out_features,
+                        bias=False,
+                        group_size=group_size,
+                        precision=precision,
+                        scales_precision=scales_precision,
+                    ),
+                )
+        else:
+            replace_linear_8da4w(
+                child,
+                group_size,
+                padding_allowed,
+                precision,
+                scales_precision,
+            )
+
+
+class Int8DynActInt4WeightGPTQQuantizer(GPTQQuantizer):
```

Comment:
Could this be made a parameter or why does it need to be a subclass?

Reply:
Or instead of making it a subclass, can it be a factory function for GPTQ?

Reply:
I think it's probably better to do it other way around, we can have `Int8DynActInt4WeightQuantizer` and have GPTQ as a separate object / addition for this quantization, as mentioned by @HDCharles